text/0000-mutex_traits.md

```diff
@@ -0,0 +1,174 @@
+
+# Detailed design
+
+## Trait, !Trait, and ?Trait
+
+An earlier RFC attempted to codify negative reasoning ran aground on the problem of backward compatibility. If it is possible to bound a type parameter or trait by the non-implementation of another trait, then that non-implementation - the _absence_ of code - becomes a semantic expression. As an example, consider this system, with traits `Consuamble`, `Edible`, and type `Dirt`.
+
+```rust
+pub trait Consumable {
+    fn consume(&self);
+}
+
+pub trait Edible {
+    fn nourish(&self);
+}
+
+pub type Dirt;
+
+impl<T> Consumable for T where T: Edible {
+    fn consume(&self) { self.nourish() }
+}
+
+impl<T> Consumable for T where T: !Edible {
+    fn consume(&self) {  }
+}
+```
+
+Under the current definitions, consuming `Dirt` has no effect, but if we were to later to discover a way to implement `Edible` for `Dirt`, that would change, and the behavior of any client relying on `Dirt`'s implementation of `Consumable` wouild change.
+
+A solution to this problem which would enable mutual exclusivity is to hold that the relation between types and traits is in one of three states for every type and trait, rather than two:
+
+* `T: Trait` - The type `T` impleemnts the trait `Trait`.
+* `T: !Trait` - The type `T` does not implement the trait `Trait`.
+* `T: ?Trait` - The type `T` may or may not implement the trait `Trait`.
+
+Without a contrary expression, the relationship between any `T` and any `Trait` is  `T: ?Trait`. This table documents the three relations and how they are described:
+
+|               | ?Trait            | Trait                | !Trait             |
+|---------------|-------------------|----------------------|--------------------|
+| Specific impl | by default        | impl Trait for T     | impl !Trait for T  |
+|               | impl ?Trait for T |                      |                    |
+| Default impl  | by default        | impl Trait for ..    | impl !Trait for .. |
+| Bounds        | by default        | where T: Trait       | where T: !Trait    |
+|               | where T: ?Sized   | by default for Sized |                    |
+
+## Defining ?Trait and !Trait
+
+`?Trait` and `!Trait` act as if they were marker traits. They define no methods and have no associated items. They are defined implicitly in the same scope as the definition of `Trait` and imported wherever `Trait` is imported.
+
+## Implementing ?Trait
+
+`?Trait` can only be implemented for types for which a default impl has been explicitly defined (e.g. `Send` and `Sync`). Explicit default impls of `?Trait` are not allowed. `?Trait` is implemented by default anyway, and it would not make sense to implement it except in the cases where an explicit default impl exists. As a rule, the syntax `?Trait` will be very uncommon.
```

Comment:
A type which implements `?Send` would not meet the bound `T: Send` or `T: !Send`. The only reason to implement `?Send` over `!Send` is if you believe you might one day be able to mark the type `Send` and don't want that to be backwards incompatible (I don't know if any types that would want this exist; I doubt that they do).

Bounding by `?Trait` is not possible and not sensible; all type parameters are bound `?Trait` for all traits unless bound otherwise, and that bound means nothing (that is, a type parameter `T` may or may not implement any particular trait unless it has a bound which declares otherwise).